server/src/datasources/talks.ts

```diff
@@ -0,0 +1,95 @@
+import {QueryBuilder} from "knex";
+import snakeCase from "lodash.snakecase";
+import {CreateTalkInput, Pagination, Talk, TalkConnection, TalkOrder, TalkSearchFilter, VoteConnection} from "../types";
+import {TalkDataSource} from "../types/datasources";
+import SelectionTree from "../types/SelectionTree";
+import SQLDataSource, {createSelectionMap, SQLDataSourceConfig} from "./SQLDataSource";
+
+
+const config: SQLDataSourceConfig = {
+    table: "talks",
+    defaultPageSize: 100,
+    relations: {
+        author: "authorId",
+    }
+}
+
+export default class TalkSQLDataSource extends SQLDataSource<Talk, CreateTalkInput> implements TalkDataSource {
+    public constructor() {
+        super(config);
+    }
+
+    public async upvote(tree: SelectionTree, id: string, by: string): Promise<Talk> {
+        await this.db.raw("INSERT IGNORE INTO talks_votes (talk_id, user_id) values (?, ?)", [id, by]);
+        await this.db.raw("UPDATE talks SET votes = (SELECT COUNT(*) FROM talks_votes WHERE id = talks.id) WHERE id = ?", id);
+        return this.load(tree, id);
+    }
+
+
+    public async search(tree: SelectionTree, filters?: TalkSearchFilter, order?: TalkOrder, page?: Pagination): Promise<TalkConnection> {
+        const query = this.table();
+
+        const totalCount = await this.getTotalCount(query, tree);
+        if (!tree.selected("nodes")) {
+            return totalCount;
+        }
+
+        const nodes = await this.getNodes(query, tree.subtree("nodes"), order, page);
+
+        return {
+            ...totalCount,
+            nodes
+        };
+    };
+
+    private async getTotalCount(query: QueryBuilder, tree: SelectionTree): Promise<TalkConnection> {
+        if (!tree.selected("totalCount")) {
+            return {} as TalkConnection;
+        }
+        query = query.clone();
+        const [row] = await query.count("* as totalCount");
```

Comment:
For a lighting talk app, the offset based pagination should quick enough to do the work.
For large scale application, offset based pagination will [go slow while the offset become large](https://www.eversql.com/faster-pagination-in-mysql-why-order-by-with-limit-and-offset-is-slow/) because the db engine will count (skip) the offset number of records which will takes time. 
The the cursor based pagination should used in place, this article shows [how to implement a cursor based pagination](https://use-the-index-luke.com/sql/partial-results/fetch-next-page).

At API design level, my current Connection objects are allow to add fields that satisfy the [Relay style cursor based API](https://relay.dev/graphql/connections.htm).